syntax = "proto3";

package auth;

service AuthService {
  rpc Login(LoginRequest) returns (LoginResponse);
  rpc Register(RegisterRequest) returns (RegisterResponse);
  rpc Logout(LogoutRequest) returns (LogoutResponse);
  
  rpc RefreshToken(RefreshTokenRequest) returns (RefreshTokenResponse);
  rpc VerifyToken(VerifyTokenRequest) returns (VerifyTokenResponse);
  rpc ForgotPassword(ForgotPasswordRequest) returns (ForgotPasswordResponse);
}

message LoginRequest {
  string email = 1;
  string password = 2;
}

message LoginResponse {
  string access_token = 1;
  string refresh_token = 2;
  string token_type = 3;
  string user_id = 4;
  int64 expires_in = 5;
  User user = 6;
}

message RegisterRequest {
  string email = 1;
  string password = 2;
  repeated string roles = 3;
}

message RegisterResponse {
  string access_token = 1;
  string refresh_token = 2;
  string token_type = 3;
  string user_id = 4;
  int64 expires_in = 5;
  User user = 6;
}

message LogoutRequest {
  string refresh_token = 1;
}

message LogoutResponse {
  bool success = 1;
  string message = 2;
}

message RefreshTokenRequest {
  string refresh_token = 1;
}

message RefreshTokenResponse {
  string access_token = 1;
  string refresh_token = 2;
  string token_type = 3;
  int64 expires_in = 4;
}

message VerifyTokenRequest {
  string token = 1;
}

message VerifyTokenResponse {
  bool valid = 1;
  string user_id = 2;
  string email = 3;
  repeated string roles = 4;
  string error = 5;
}

message ForgotPasswordRequest {
  string email = 1;
}

message ForgotPasswordResponse {
  bool success = 1;
  string reset_token = 2;
  string message = 3;
}

message User {
  string id = 1;
  string email = 2;
  repeated string roles = 3;
  bool is_banned = 4;
  string created_at = 5;
  string last_login_at = 6;
}